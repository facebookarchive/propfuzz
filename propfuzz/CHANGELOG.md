# Changelog

## [0.0.1] - 2020-07-01
### Added
- Initial release.

[0.0.1]: https://github.com/facebookincubator/cargo-guppy/releases/tag/propfuzz-0.0.1
