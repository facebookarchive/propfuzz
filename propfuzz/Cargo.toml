[package]
name = "propfuzz"
version = "0.0.1"
description = "Combine property-based testing and fuzzing."
repository = "https://github.com/facebookincubator/propfuzz"
authors = ["Rain <rain1@fb.com>"]
license = "MIT OR Apache-2.0"
edition = "2018"
readme = "README.md"
exclude = [
    # Readme template that doesn't need to be included.
    "README.tpl",
]

[badges]
maintenance = { status = "actively-developed" }

[package.metadata.docs.rs]
all-features = true

[dependencies]
proptest = "1.0.0"
propfuzz-macro = { path = "../propfuzz-macro", version = "0.0.1", optional = true }

[dev-dependencies]
pretty_assertions = "1.0.0"
trybuild = "1.0.53"

[features]
default = ["macro"]
macro = ["propfuzz-macro"]

[[test]]
name = "basic"
required-features = ["macro"]

[[test]]
name = "trybuild"
required-features = ["macro"]
