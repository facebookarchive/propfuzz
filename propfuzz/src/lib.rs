// Copyright (c) The propfuzz Contributors
// SPDX-License-Identifier: MIT OR Apache-2.0

//! Combine property-based testing with fuzzing.

mod propfuzz;
pub mod runtime;

pub use crate::propfuzz::*;

// Re-export proptest to give access to its data structures.
#[doc(no_inline)]
pub use proptest as test;
