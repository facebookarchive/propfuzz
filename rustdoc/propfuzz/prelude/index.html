<!DOCTYPE html><html lang="en"><head><meta charset="utf-8"><meta name="viewport" content="width=device-width, initial-scale=1.0"><meta name="generator" content="rustdoc"><meta name="description" content="API documentation for the Rust `prelude` mod in crate `propfuzz`."><meta name="keywords" content="rust, rustlang, rust-lang, prelude"><title>propfuzz::prelude - Rust</title><link rel="stylesheet" type="text/css" href="../../normalize.css"><link rel="stylesheet" type="text/css" href="../../rustdoc.css" id="mainThemeStyle"><link rel="stylesheet" type="text/css" href="../../light.css"  id="themeStyle"><link rel="stylesheet" type="text/css" href="../../dark.css" disabled ><link rel="stylesheet" type="text/css" href="../../ayu.css" disabled ><script id="default-settings"></script><script src="../../storage.js"></script><noscript><link rel="stylesheet" href="../../noscript.css"></noscript><link rel="icon" type="image/svg+xml" href="../../favicon.svg">
<link rel="alternate icon" type="image/png" href="../../favicon-16x16.png">
<link rel="alternate icon" type="image/png" href="../../favicon-32x32.png"><style type="text/css">#crate-search{background-image:url("../../down-arrow.svg");}</style></head><body class="rustdoc mod"><!--[if lte IE 8]><div class="warning">This old browser is unsupported and will most likely display funky things.</div><![endif]--><nav class="sidebar"><div class="sidebar-menu">&#9776;</div><a href='../../propfuzz/index.html'><div class='logo-container rust-logo'><img src='../../rust-logo.png' alt='logo'></div></a><p class="location">Module prelude</p><div class="sidebar-elems"><div class="block items"><ul><li><a href="#reexports">Re-exports</a></li></ul></div><p class="location"><a href="../index.html">propfuzz</a></p><script>window.sidebarCurrent = {name: "prelude", ty: "mod", relpath: "../"};</script><script defer src="../sidebar-items.js"></script></div></nav><div class="theme-picker"><button id="theme-picker" aria-label="Pick another theme!" aria-haspopup="menu"><img src="../../brush.svg" width="18" alt="Pick another theme!"></button><div id="theme-choices" role="menu"></div></div><script src="../../theme.js"></script><nav class="sub"><form class="search-form"><div class="search-container"><div><select id="crate-search"><option value="All crates">All crates</option></select><input class="search-input" name="search" disabled autocomplete="off" spellcheck="false" placeholder="Click or press ‘S’ to search, ‘?’ for more options…" type="search"></div><span class="help-button">?</span>
                <a id="settings-menu" href="../../settings.html"><img src="../../wheel.svg" width="18" alt="Change settings"></a></div></form></nav><section id="main" class="content"><h1 class="fqn"><span class="out-of-band"><span id="render-detail"><a id="toggle-all-docs" href="javascript:void(0)" title="collapse all docs">[<span class="inner">&#x2212;</span>]</a></span><a class="srclink" href="../../src/propfuzz/prelude.rs.html#4-31" title="goto source code">[src]</a></span><span class="in-band">Module <a href="../index.html">propfuzz</a>::<wbr><a class="mod" href="">prelude</a></span></h1><div class="docblock"><p>Re-exports for the most commonly used APIs of <code>propfuzz</code>.</p>
<p>This includes:</p>
<ul>
<li>the <code>propfuzz</code> macro from this crate</li>
<li>the entire prelude of <code>proptest</code>, so existing tests can be migrated with minimal hassle.</li>
</ul>
<h2 id="examples" class="section-header"><a href="#examples">Examples</a></h2>
<p>There's no need to specify <code>proptest</code> as a separate dependency, since you can write:</p>

<div class="example-wrap"><pre class="rust rust-example-rendered">
<span class="kw">use</span> <span class="ident">propfuzz</span>::<span class="ident">prelude</span>::<span class="kw-2">*</span>;
<span class="kw">use</span> <span class="ident">proptest</span>::<span class="ident">collection</span>::<span class="ident">vec</span>;

<span class="doccomment">/// Example test.</span>
<span class="attribute">#[<span class="ident">propfuzz</span>]</span>
<span class="kw">fn</span> <span class="ident">test</span>(<span class="attribute">#[<span class="ident">propfuzz</span>(<span class="ident">strategy</span> <span class="op">=</span> <span class="string">&quot;vec(any::&lt;u8&gt;(), 0..64)&quot;</span>)]</span> <span class="ident">v</span>: <span class="ident">Vec</span><span class="op">&lt;</span><span class="ident">u8</span><span class="op">&gt;</span>) {
    <span class="comment">// ...</span>
}</pre></div>
</div><h2 id="reexports" class="section-header"><a href="#reexports">Re-exports</a></h2>
<table><tr><td><code>pub use crate::<a class="attr" href="../../propfuzz/attr.propfuzz.html" title="attr propfuzz::propfuzz">propfuzz</a>;</code></td></tr><tr><td><code>pub use <a class="mod" href="../../proptest/index.html" title="mod proptest">proptest</a>;</code></td></tr><tr><td><code>pub use <a class="macro" href="../../proptest/macro.proptest.html" title="macro proptest::proptest">proptest</a>;</code></td></tr><tr><td><code>pub use proptest::<a class="mod" href="../../proptest/prelude/index.html" title="mod proptest::prelude">prelude</a>::*;</code></td></tr></table></section><section id="search" class="content hidden"></section><section class="footer"></section><script>window.rootPath = "../../";window.currentCrate = "propfuzz";</script><script src="../../main.js"></script><script defer src="../../search-index.js"></script></body></html>