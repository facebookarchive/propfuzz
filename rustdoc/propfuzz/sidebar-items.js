initSidebarItems({"attr":[["propfuzz","The core macro, used to annotate test methods."]],"mod":[["prelude","Re-exports for the most commonly used APIs of `propfuzz`."],["runtime","Runtime support."],["traits","The core traits powering `propfuzz`."]]});