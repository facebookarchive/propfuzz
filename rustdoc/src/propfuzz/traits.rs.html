<!DOCTYPE html><html lang="en"><head><meta charset="utf-8"><meta name="viewport" content="width=device-width, initial-scale=1.0"><meta name="generator" content="rustdoc"><meta name="description" content="Source of the Rust file `propfuzz/src/traits.rs`."><meta name="keywords" content="rust, rustlang, rust-lang"><title>traits.rs - source</title><link rel="stylesheet" type="text/css" href="../../normalize.css"><link rel="stylesheet" type="text/css" href="../../rustdoc.css" id="mainThemeStyle"><link rel="stylesheet" type="text/css" href="../../light.css"  id="themeStyle"><link rel="stylesheet" type="text/css" href="../../dark.css" disabled ><link rel="stylesheet" type="text/css" href="../../ayu.css" disabled ><script id="default-settings"></script><script src="../../storage.js"></script><noscript><link rel="stylesheet" href="../../noscript.css"></noscript><link rel="icon" type="image/svg+xml" href="../../favicon.svg">
<link rel="alternate icon" type="image/png" href="../../favicon-16x16.png">
<link rel="alternate icon" type="image/png" href="../../favicon-32x32.png"><style type="text/css">#crate-search{background-image:url("../../down-arrow.svg");}</style></head><body class="rustdoc source"><!--[if lte IE 8]><div class="warning">This old browser is unsupported and will most likely display funky things.</div><![endif]--><nav class="sidebar"><div class="sidebar-menu">&#9776;</div><a href='../../propfuzz/index.html'><div class='logo-container rust-logo'><img src='../../rust-logo.png' alt='logo'></div></a></nav><div class="theme-picker"><button id="theme-picker" aria-label="Pick another theme!" aria-haspopup="menu"><img src="../../brush.svg" width="18" alt="Pick another theme!"></button><div id="theme-choices" role="menu"></div></div><script src="../../theme.js"></script><nav class="sub"><form class="search-form"><div class="search-container"><div><select id="crate-search"><option value="All crates">All crates</option></select><input class="search-input" name="search" disabled autocomplete="off" spellcheck="false" placeholder="Click or press ‘S’ to search, ‘?’ for more options…" type="search"></div><span class="help-button">?</span>
                <a id="settings-menu" href="../../settings.html"><img src="../../wheel.svg" width="18" alt="Change settings"></a></div></form></nav><section id="main" class="content"><pre class="line-numbers"><span id="1"> 1</span>
<span id="2"> 2</span>
<span id="3"> 3</span>
<span id="4"> 4</span>
<span id="5"> 5</span>
<span id="6"> 6</span>
<span id="7"> 7</span>
<span id="8"> 8</span>
<span id="9"> 9</span>
<span id="10">10</span>
<span id="11">11</span>
<span id="12">12</span>
<span id="13">13</span>
<span id="14">14</span>
<span id="15">15</span>
<span id="16">16</span>
<span id="17">17</span>
<span id="18">18</span>
<span id="19">19</span>
<span id="20">20</span>
<span id="21">21</span>
<span id="22">22</span>
<span id="23">23</span>
<span id="24">24</span>
<span id="25">25</span>
<span id="26">26</span>
<span id="27">27</span>
<span id="28">28</span>
<span id="29">29</span>
<span id="30">30</span>
<span id="31">31</span>
<span id="32">32</span>
<span id="33">33</span>
<span id="34">34</span>
<span id="35">35</span>
<span id="36">36</span>
<span id="37">37</span>
<span id="38">38</span>
<span id="39">39</span>
<span id="40">40</span>
<span id="41">41</span>
<span id="42">42</span>
<span id="43">43</span>
<span id="44">44</span>
<span id="45">45</span>
<span id="46">46</span>
<span id="47">47</span>
<span id="48">48</span>
<span id="49">49</span>
<span id="50">50</span>
<span id="51">51</span>
<span id="52">52</span>
<span id="53">53</span>
<span id="54">54</span>
<span id="55">55</span>
<span id="56">56</span>
</pre><div class="example-wrap"><pre class="rust ">
<span class="comment">// Copyright (c) The propfuzz Contributors</span>
<span class="comment">// SPDX-License-Identifier: MIT OR Apache-2.0</span>

<span class="doccomment">//! The core traits powering `propfuzz`.</span>
<span class="doccomment">//!</span>
<span class="doccomment">//! `propfuzz` relies on separating out a standard property-based test into its components:</span>
<span class="doccomment">//! * constructing a configuration</span>
<span class="doccomment">//! * executing the test, given a test runner</span>
<span class="doccomment">//! * formatting failing values</span>

<span class="kw">use</span> <span class="ident">proptest</span>::<span class="ident">prelude</span>::<span class="kw-2">*</span>;
<span class="kw">use</span> <span class="ident">proptest</span>::<span class="ident">test_runner</span>::{<span class="ident">TestError</span>, <span class="ident">TestRunner</span>};
<span class="kw">use</span> <span class="ident">std</span>::<span class="ident">fmt</span>;

<span class="doccomment">/// Represents a structured fuzz target.</span>
<span class="doccomment">///</span>
<span class="doccomment">/// Functions written with the `propfuzz` macro are converted to implementations of this trait.</span>
<span class="doccomment">///</span>
<span class="doccomment">/// A trait that implements `Propfuzz` can be used both as a standard property-based test, and</span>
<span class="doccomment">/// as a target for structured, mutation-based fuzzing.</span>
<span class="doccomment">///</span>
<span class="doccomment">/// Structured, mutation-based fuzzers use random byte sequences as a pass-through RNG. In other</span>
<span class="doccomment">/// words, the random byte sequences act as something like a DNA for random values.</span>
<span class="kw">pub</span> <span class="kw">trait</span> <span class="ident">StructuredTarget</span>: <span class="ident">Send</span> <span class="op">+</span> <span class="ident">Sync</span> <span class="op">+</span> <span class="ident">fmt</span>::<span class="ident">Debug</span> {
    <span class="doccomment">/// The type of values generated by the proptest.</span>
    <span class="kw">type</span> <span class="ident">Value</span>: <span class="ident">fmt</span>::<span class="ident">Debug</span>;

    <span class="doccomment">/// Returns the name of this structured fuzz target.</span>
    <span class="kw">fn</span> <span class="ident">name</span>(<span class="kw-2">&amp;</span><span class="self">self</span>) <span class="op">-</span><span class="op">&gt;</span> <span class="kw-2">&amp;</span><span class="lifetime">&#39;static</span> <span class="ident">str</span>;

    <span class="doccomment">/// Returns an optional description for this structured fuzz target.</span>
    <span class="kw">fn</span> <span class="ident">description</span>(<span class="kw-2">&amp;</span><span class="self">self</span>) <span class="op">-</span><span class="op">&gt;</span> <span class="prelude-ty">Option</span><span class="op">&lt;</span><span class="kw-2">&amp;</span><span class="lifetime">&#39;static</span> <span class="ident">str</span><span class="op">&gt;</span>;

    <span class="doccomment">/// Returns the proptest config for this fuzz target.</span>
    <span class="doccomment">///</span>
    <span class="doccomment">/// The default implementation for the `#[propfuzz]` macro uses the default `proptest` config,</span>
    <span class="doccomment">/// and sets `source_file` to the file the macro is invoked from. The config can be altered</span>
    <span class="doccomment">/// through passing arguments to `#[propfuzz]`.</span>
    <span class="kw">fn</span> <span class="ident">proptest_config</span>(<span class="kw-2">&amp;</span><span class="self">self</span>) <span class="op">-</span><span class="op">&gt;</span> <span class="ident">ProptestConfig</span> {
        <span class="ident">ProptestConfig</span>::<span class="ident">default</span>()
    }

    <span class="doccomment">/// Executes this test using the given test runner.</span>
    <span class="doccomment">///</span>
    <span class="doccomment">/// This is where the main body of the test goes.</span>
    <span class="kw">fn</span> <span class="ident">execute</span>(<span class="kw-2">&amp;</span><span class="self">self</span>, <span class="ident">test_runner</span>: <span class="kw-2">&amp;</span><span class="kw-2">mut</span> <span class="ident">TestRunner</span>) <span class="op">-</span><span class="op">&gt;</span> <span class="prelude-ty">Result</span><span class="op">&lt;</span>(), <span class="ident">TestError</span><span class="op">&lt;</span><span class="self">Self</span>::<span class="ident">Value</span><span class="op">&gt;</span><span class="op">&gt;</span>;

    <span class="doccomment">/// Formats a failing test case for displaying to the user.</span>
    <span class="doccomment">///</span>
    <span class="doccomment">/// The default implementation calls the `fmt::Debug` implementation on the value.</span>
    <span class="doccomment">///</span>
    <span class="doccomment">/// The default implementation for the `#[propfuzz]` macro prints out variable names and values.</span>
    <span class="kw">fn</span> <span class="ident">fmt_value</span>(<span class="kw-2">&amp;</span><span class="self">self</span>, <span class="ident">value</span>: <span class="kw-2">&amp;</span><span class="self">Self</span>::<span class="ident">Value</span>, <span class="ident">f</span>: <span class="kw-2">&amp;</span><span class="kw-2">mut</span> <span class="ident">fmt</span>::<span class="ident">Formatter</span>) <span class="op">-</span><span class="op">&gt;</span> <span class="ident">fmt</span>::<span class="prelude-ty">Result</span> {
        <span class="macro">write</span><span class="macro">!</span>(<span class="ident">f</span>, <span class="string">&quot;{:?}&quot;</span>, <span class="ident">value</span>)
    }
}
</pre></div>
</section><section id="search" class="content hidden"></section><section class="footer"></section><script>window.rootPath = "../../";window.currentCrate = "propfuzz";</script><script src="../../main.js"></script><script src="../../source-script.js"></script><script src="../../source-files.js"></script><script defer src="../../search-index.js"></script></body></html>