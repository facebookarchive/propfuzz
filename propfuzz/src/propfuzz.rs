// Copyright (c) The propfuzz Contributors
// SPDX-License-Identifier: MIT OR Apache-2.0

use proptest::prelude::*;
use proptest::test_runner::TestRunner;
use std::fmt;

/// Represents a structured fuzz target.
///
/// A trait that implements `Propfuzz` can be used both as a standard property-based test, and
/// as a target for structured, mutation-based fuzzing.
///
/// Structured, mutation-based fuzzers use random byte sequences as a pass-through RNG. In other
/// words, the random byte sequences act as something like a DNA for random values.
pub trait Propfuzz: Send + Sync + fmt::Debug {
    /// Returns the name of this structured fuzz target.
    fn name(&self) -> &'static str;

    /// Returns a description for this structured fuzz target.
    fn description(&self) -> &'static str;

    /// Returns the proptest config for this fuzz target.
    ///
    /// Note that the config is modified for the
    fn proptest_config(&self) -> ProptestConfig {
        ProptestConfig::default()
    }

    /// Executes this test using the given test runner.
    ///
    /// This is where the main body of the test goes.
    fn execute(&self, test_runner: &mut TestRunner);
}
