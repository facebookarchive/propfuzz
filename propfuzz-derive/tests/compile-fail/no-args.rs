// Copyright (c) The propfuzz Contributors
// SPDX-License-Identifier: MIT OR Apache-2.0

//! A propfuzz with no arguments.

use propfuzz_derive::propfuzz;

#[propfuzz]
fn no_args() {}

fn main() {}
