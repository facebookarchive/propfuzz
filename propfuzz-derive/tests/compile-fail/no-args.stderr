error: #[propfuzz] requires at least one argument
 --> $DIR/no-args.rs:9:1
  |
9 | fn no_args() {}
  | ^^^^^^^^^^^^
